compressors/typical_set_coder.py

```diff
@@ -0,0 +1,208 @@
+"""Typical set coder
+
+For a given iid source $X_1,X_2,...$ i.i.d. $\sim X$ (from alphabet $\mathcal{X}$), 
+and for given values of $n > 0$ and $\epsilon > 0$, the typical set is defined as 
+$$A_{\epsilon}^{(n)} = \left\{x^n | \left|-\frac{1}{n}\log P(x^n) - H(X)\right| \leq \epsilon\right\}$$
+
+This code works by dividing the input into blocks of size n, and then encodes each block
+as follows (note: a single data_block contains several of these n-length blocks):
+- if $x^n \in A_{\epsilon}^{(n)}$, encode as a $0$ followed by a 
+$\lceil\log_2|A_{\epsilon}^{(n)}|\rceil$ bit representation of the index of $x^n$ in the typical set
+- otherwise, encode as $1$ followed by a $\lceil\log_2|\mathcal{X}|\rceil$ representation 
+of the index of $x^n$ in $\mathcal{X}^n$
+"""
+
+from core.data_block import DataBlock
+from core.data_encoder_decoder import DataDecoder, DataEncoder
+from utils.bitarray_utils import BitArray, uint_to_bitarray, bitarray_to_uint
+from utils.test_utils import get_random_data_block, try_lossless_compression
+from core.prob_dist import ProbabilityDist
+import numpy as np
+import itertools
+import random
+
+
+def compute_normalized_negative_log_prob_block(block, prob_dist: ProbabilityDist):
+    """Compute the normalized log probability for a block"""
+    log_prob = 0
+    for symbol in block:
+        log_prob += -np.log2(prob_dist.probability(symbol))
+    return log_prob / len(block)
+
+
+def is_typical(block, prob_dist: ProbabilityDist, eps):
+    """Check if a block is typical given a distribution and value of epsilon"""
+    if np.abs(compute_normalized_negative_log_prob_block(block, prob_dist) - prob_dist.entropy) <= eps:
+        return True
+    else:
+        return False
+
+
+def generate_indexes(prob_dist, n, eps):
+    """generate the indexes used for encoding/decoding of typical set code"""
+    index_typical = {}  # mapping from input typical sequence to its index in the typical set
+    index_overall = {}  # mapping from an input sequence to its index in \mathcal{X}^n set
+    # both of these map tuples of length n to non-negative integers
+
+    num_typical_seen = 0  # use for indexing typical
+    num_overall_seen = 0  # use for indexing overall
+
+    for l in itertools.product(prob_dist.alphabet, repeat=n):
+        if is_typical(l, prob_dist, eps):
+            index_typical[tuple(l)] = num_typical_seen
+            num_typical_seen += 1
+        index_overall[tuple(l)] = num_overall_seen
+        num_overall_seen += 1
+
+    return index_typical, index_overall
+
+
+class TypicalSetEncoder(DataEncoder):
+    """Typical set encoder working on blocks of length n. See module documentation for more details."""
+
+    def __init__(self, prob_dist, n, eps) -> None:
```

Comment:
1. please add type hints for `n, eps` (`n: int, eps: float`) 
2. Please add a comment for `n`, `eps` as these are not descriptive var names (both here and in the decoder)

Reply:
Another thing which might make this very clear is creating a class:
```
@dataclass
class TypicalSetParams:
    # add more comments
     n: int = None #the chunk size
     eps: float: = 0 # the tolerance.. 

```

this might make it simpler and explicit that `n,eps` are typical set params?

Reply:
done